### DISCUSSION TITLE
Is it possible to ignore self-signed errors using `fetch`?

### DISCUSSION THREAD

**Discussion by astrodomas** [vercel/next.js#74187] [App Router]
Hello, since nextjs mandates using the web fetch api which it extends. Is it possible to ignore self-signed tls certs? As to my knowledge web fetch does not allow that. Meaning, that on the server-side I can't call my api layer.

Using nextjs built in fetch - there is no option to pass a httpsAgent or anything, that would give an option to ignore the TLS.

![image](https://github.com/user-attachments/assets/7dec7123-13c7-4b02-87d0-97e523a0d355)
![image](https://github.com/user-attachments/assets/c62065df-736b-4d8f-b5f6-d2558d2350d1)

Also, this is probably not related to this issue (as it's regarding the nextjs middleware layer): https://github.com/vercel/next.js/discussions/49546

Code snippets of working examples

```
const fetch = require('node-fetch');
const https = require('https');

const { request, Agent, ProxyAgent } = require('undici');

const httpsAgent = new https.Agent({
  rejectUnauthorized: false,
});

const node_fetch_req = async () => {
  const response = await fetch('https://mySecureApi/api/v1/endpoint', {
    headers: {
      Host: 'hostname',
    },
    agent: httpsAgent,
  })
    .then((response) => {
      // Check if the response is JSON
      if (response.headers.get('Content-Type')?.includes('application/json')) {
        return response.json(); // Parse the JSON
      } else {
        return response.text(); // Otherwise, return the response as text (could be HTML)
      }
    })
    .then((data) => {
      // Handle the response (either JSON or HTML)
      console.log(data);
    });
  return response;
};

node_fetch_req();

const undici_fetch_req = async () => {
  const response = await request('https://mySecureApi/api/v1/endpoint', {
    headers: {
      Host: 'hostname',
    },
    dispatcher: new Agent({
      connect: {
        rejectUnauthorized: false,
      },
    }),
  })
    .then(async (resp) => {
      const content = await resp.body.text();
      console.log(content);
    })
    .catch((error) => {
      console.error('Error:', error);
    });

  return response;
};

undici_fetch_req();

const native_fetch_request = async () => {
  const response = await fetch('https://mySecureApi/api/v1/endpoint', {
    headers: {
      Host: 'hostname',
    },
    dispatcher: new Agent({
      connect: {
        rejectUnauthorized: false,
      },
    }),
  })
    .then(async (resp) => {
      const content = await resp.text();
      console.log(content);
    })
    .catch((error) => {
      console.error('Error:', error);
    });

  return response;
};

native_fetch_request();



```

**Comment by icyJoseph** [comment]
Could you read through here? https://github.com/nodejs/undici/issues/1489#issuecomment-1543856261

**Reply by astrodomas** [reply]
Thanks for the answer, However I already tried this approach and it does not work. As you can see in my codesnippet
![image](https://github.com/user-attachments/assets/95903a99-5921-4944-87d8-6e074a62644e)
`dispatcher` option is not available

**Reply by astrodomas** [reply]
if I try it with `import { request } from 'undici'`, then it works, however it's not an option to use undici.

**Reply by icyJoseph** [reply]
When you try with the dispatcher, does it not work at all? or is it just a TypeScript error?

- https://github.com/nodejs/undici/discussions/2167#discussioncomment-6239415

This seems to suggest that one should be able to override this, one way or another.

**Reply by astrodomas** [reply]
Yeah, I'm getting `404`, while getting the desired api response (200 with json payloed) on the working snippets provided. I also have tried setting the global dispatcher

**Reply by astrodomas** [reply]
I believe nextjs uses undici under the hood?

**Reply by icyJoseph** [reply]
Nope, Node.js uses undici, that's the implementation they went with.

Next.js does some patching but the agent options should work 🤔 

What Node.js version have you got?

Could you make a simple Node.js script with these changes and see if you can get it to work there? That is to say, let's remove Next.js from the equation, and see if a simple `node index.js` works? Just do `yarn init -y`, or the like, install undici, and yeah, not so simple I guess, but it should take you no time 🙏

**Reply by astrodomas** [reply]
The provided code snippets are without nextjs and they work, however I see that I used undici `request`.
I will upload a snippet that uses undici `fetch` shortly

**Reply by astrodomas** [reply]
example test.js using `fetch` from undici. Getting 404, however with other examples (using `node-fetch` and undici `request` I get the desired output)
```
const { fetch: undici_fetch } = require('undici');

const undici_fetch_request = async () => {
  const response = await undici_fetch('https://mySecureApi/api/v1/endpoint', {
    headers: {
      Host: 'hostname',
    },
    dispatcher: new Agent({
      connect: {
        rejectUnauthorized: false,
      },
    }),
  })
    .then(async (resp) => {
      const content = await resp.text();
      console.log(content);
    })
    .catch((error) => {
      console.error('Error:', error);
    });

  return response;
};

undici_fetch_request();

```


```
<html>
<head><title>404 Not Found</title></head>
<body>
<body>
<center><h1>404 Not Found</h1></center>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
<hr><center>nginx</center>
</body>
</html>
```

**Reply by icyJoseph** [reply]
But you should just use the global fetch, no need to take it from undici, not sure that matters though. Node.js supports fetch globally since 18 or so

**Reply by astrodomas** [reply]
```
node -v
v20.9.0
```

**Reply by icyJoseph** [reply]
I wonder if you have a Node.js version without global fetch? And then Next.js polyfills with node-fetch? I can't remember if they still do that, Next.js require Node 18 as well.

**Reply by astrodomas** [reply]
I'll try the same fetch with node 18

**Reply by icyJoseph** [reply]
That should've been a safe version. Ugh, this kind of thing shouldn't be so complicated 😅

I have ran out of time for now. https://github.com/vercel/next.js/blob/canary/packages%2Fnext%2Fsrc%2Fserver%2Flib%2Fpatch-fetch.ts here's where Next.js patches the global fetch. 

Do keep posting updates, maybe later today I can jump back in.

**Reply by astrodomas** [reply]
Tried with node v18 and v20, and native fetch, same thing. Getting 404:

```
const native_fetch_request = async () => {
  const response = await fetch('https://mySecureApi/api/v1/endpoint', {
    headers: {
      Host: 'hostname',
    },
    dispatcher: new Agent({
      connect: {
        rejectUnauthorized: false,
      },
    }),
  })
    .then(async (resp) => {
      const content = await resp.text();
      console.log(content);
    })
    .catch((error) => {
      console.error('Error:', error);
    });

  return response;
};

native_fetch_request();
```

**Reply by icyJoseph** [reply]
Mmm maybe the agent needs more options? Super odd, of all errors a 404 is kind of weird 

Are you able to debug this on the other API?

Fetching pokemon, todos, etc other data, does work right? (With a custom agent)

**Reply by astrodomas** [reply]
I have change the `dispatcher` property to `agent`. Seems to be working on https://google.com, however not on my nginx server:
    `code: 'ERR_TLS_CERT_ALTNAME_INVALID'` . I guess this will be related to my nginx server.

**Comment by e23thr** [comment]
May this help. Setting environment variable "NODE_TLS_REJECT_UNAUTHORIZED" to "0"

```
process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = "0";
```

**Reply by astrodomas** [reply]
https://nodejs.org/download/release/v10.16.0/docs/api/cli.html
```
If value equals '0', certificate validation is disabled for TLS connections. This makes TLS, and HTTPS by extension, insecure. The use of this environment variable is strongly discouraged.
```

My usecase is granular, not global.
